content/blog/vue-translate.md

```diff
@@ -0,0 +1,175 @@
+            inputType="text"
+            name="myEmailInput"
+            value="">
+</translated-input>
+```
+
+But unlike React and more like Angular, Vue is also a framework of directives. A directive looks 
+like a custom attribute on an HTML element and defines some behaviour on that element. It would be much
+less intrusive if we could define a custom directive that would translate any part of an existing element.
+The desired interface would be:
+
+```
+<h3 v-translate="'welcomeMessage'"></h3>
+<input type="text" v-translate:placeholder="'email'" name="myEmailInput" value="" />
+```
+
+This is clearly a much nicer interface, but is not as straight-forward to implement. This blog will describe
+how we did it!
+
+### Storing and updating the user's chosen language
+In Vuex all app state is held
+in a central "store", which is then the ideal place to keep track of what language the user has selected.
+
+```
+const store = new Vuex.Store({
+    state: {
+        language: "en     
+    } 
+})
+```
+
+For performing the translations we have used [i18next](https://www.i18next.com/), which is intialised
+ with a language and the available translations. When the user selects a new language we fire off a Vuex
+  action that updates both i18next and the language in our store state.
+ 
+ ```
+ await i18next.changeLanguage(newLanguage);
+ commit({type: "CHANGE_LANGUAGE", payload: newLanguage})
+```
+
+### Listening for changes
+
+Listening for changes in a component is easy, it's baked into the framework. A simple translation component
+that just translates text and renders it in a span looks like this:
+
+```
+<template>
+    <span>{{translatedText}}</span>
+</template>
+
+<script>
+import {mapState} from Vuex
+
+export default {
+    props: ["translationKey"],
+    data() {
+        return {
+            translatedText: i18next.t(this.translationKey)
+        }
+    },
+    computed: mapState({ language -> state.language }),
+    watch: {
+        language() {
+            this.translatedText = i18next.t(this.translationKey)
+        }
+    }
+}
+
+</script>
+```
+
+The computed property "language" gets updated whenever the store state language changes, the watcher fires and the 
+text is updated.
+
+A directive has a completely different structure. There are 5 lifecycle events: `bind`, `inserted`, `update`, 
+`componentUpdated` and `unbind`. `bind` is fired when the component is first created, and the `update` hook allows us to listen for changes to the value of the directive. 
+It will fire if for example, the prop "title" changes in value in this component:
+
+```
+<template>
+    <h3 v-translate="title"></h3>
+</template>
+<script>
+export default {
+    props: ["title"]
+}
+</script>
+```
+
+But there is no directive hook that will fire when *store state* changes, i.e. our language. So we'll have to manage
+a subscription to the store ourselves. We can set up such a subscription on `bind`:
+
+```
+bind(el, binding) {
```

Comment:
Maybe put in more of an explanation of the getting and using of the unwatch function since that confused me!